sources/appendix-value-extraction.asciidoc

```diff
@@ -0,0 +1,437 @@
+* Validation of type-level constraints as enabled by Java 8: `Optional<@Email String> email` or `Property<@Min(1) Integer> value`; in this case, e.g. the wrapped `String` and `Integer` values must be extracted so the `@Email` and `@Min` constraints can be applied
+
+Both cases can also overlap: `List<@RetailOrder @Valid> retailOrders`.
+
+Type-level constraints and pluggable extractors will also make cascaded validation more flexible. As of BV 1.1, there was a fixed set of data types supported by cascaded validation mandated by the spec (`Iterable`, arrays, `Map` (only the values would be validated)). This excludes use cases such as custom collection types (e.g. Guava's multi-map), validation of map keys or collection types of other JVM languages (such as Ceylon's collection framework).
+
+Allowing to put `@Valid` on type parameters allows to express the subject of cascaded validation more specifically: `Map<@Valid AddressType, @Valid Address> addresses` would describe that both, map keys and values, should be validated.
+
+=== Requirements
+
+1. Constraints can be applied to type parameters (of bean properties, but also method parameters etc.): `Property<@Min(1) Integer> value`
+1. Individual type parameters can be marked for cascaded validation: `Map<@Valid AddressType, Address> addresses`
+1. How values are retrieved must be customizable by means of a value extractor SPI, with a set of defined default extractors
+1. Sometimes constraints should apply to a value wrapped by a container type, but there is no type parameter to put the constraints to. JavaFX's `Property` hierarchy is the most prominent example: `@Email StringProperty emailProperty`. In such case the value should be implicitly extracted from the container if unambiguously doable.
+1. OPEN: Support for validation of nested collections: `List<Map<String, @NotNull String>> addressesByType`.
+1. Allow to configure type-level constraints and cascades via XML mappings
+1. Expose meta-data on type-level constraints and cascades in the constraint metadata API
+
+=== Non-requirements
+
+1. Constraints on type uses in local variables, invocations etc: `@NotNull String name = "Emmanuel";`, `new @NonEmpty @Readonly List<String>(myNonEmptyStringSet)`
+1. constraints on type uses in type definitions:
+
+[source,java]
+----
+class CustList extends List<@NotNull Customer> {
+}
+----
+
+=== Solution
+
+Value retrieval for cascaded validation and validation of type-level constraints is done via the `ValueExtractor` SPI:
+
+[source,java]
+----
+package javax.validation.valueretrieval;
+
+public interface ValueExtractor<T> {
+
+	void extractValues(T originalValue, ValueReceiver receiver);
+
+	interface ValueReceiver {
+
+		void value(Object object, String nodeName);
+
+		void iterableValue(Object object, String nodeName);
```

Comment:
@gunnarmorling I was thinking of offering a contract to convert an object into a `Path.node`

```
class  GuavaTableValueExtractor implements ValueExtractor {
    void extractValues(GuavaTable originalValue, ValueReceiver receiver);
        Set< Table.Cell<R, C, V> > cells = originalValue.cellSet()
        for (Table.Cell cell: cells) {
            receiver.valueWithKey( cell.getValue(), cell );
        }

    }
}

class CellToKeyPathNodeConverter implements PathNodeConverter<Table.Cell> {
    NodeKey convert(TableCell cell) {
        return new NodeKey() {
              String getName() {
                  return "cell.getRow() + ", " + cell.getColumn();
                   // TODO what is the prefix, ideally I'd like a path like tables[row1,column1]
              }
    }
}
```

But thinking about it more:

* it ends up being just a string reprensentation so maybe the string param is enough
* sharing implementations of object to `Path.Node` ends up ambiguous (several version of the same object type might have different representations) so putting the impl in the `ValueExtractor` is probably better

But I'm still confused as to what the nodes would look like both in string and in actual `Path.Node` approach. Examples would help.